### PR TITLE
[DONOTMERGE] Support a CNCI per subnet per tenant in ciao-controller

### PR DESCRIPTION
**Pull request by kaccardi**
This PR implements multiple CNCIs per tenant in the controller, one for each subnet. When a subnet is no longer in use, the CNCI for that subnet will be deleted after a 5 minute settle period. It creates a CNCI controller interface per tenant. This controller is responsible for managing the CNCI lifecycle.  With this code, we will now store CNCI instance information in the datastore with the tenant instances so that we can delete CNCI instances when we no longer need them.

### CODE REVIEW

File: ciao-controller/client.go [hunk 1]
```diff
@@ -162,6 +179,25 @@ func (client *ssntpClient) instanceAdded(payload []byte) {
...
+	}
+
+	tenant, err := client.ctl.ds.GetTenant(i.TenantID)
+	if err != nil || tenant == nil {
+		glog.Warningf("Error getting tenant: %v", err)
+		return
+	}
+
+	if tenant.CNCIctrl == nil {
+		fmt.Println("No CNCI controller saved")
```
Comment (kaccardi) [inline ciao-controller/client.go]: stray debug left in here.

File: ciao-controller/cnci.go [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+	cnci, ok := c.cncis[ID]
+	if !ok {
+		return false
+	}
+
+	return cnci.active()
+}
+
+func (c *CNCIManager) launch() (string, error) {
+	glog.Info("launching cnci")
```
Comment (kaccardi) [inline ciao-controller/cnci.go]: remove stray debug

File: ciao-controller/internal/datastore/datastore.go [hunk 3]
```diff
@@ -772,6 +795,24 @@ func (ds *Datastore) AllocateTenantIP(tenantID string) (net.IP, error) {
...
+		err := mgr.AddSubnet(int(subnetInt))
+		if err != nil {
+			glog.Warningf("Error adding subnet (%v)\n", err)
+		}
+	}
+
+	// wait for cnci to become ready before moving on.
+	if mgr != nil {
+		err := mgr.WaitForActive(int(subnetInt))
+		if err != nil {
```
Comment (rbradford) [inline ciao-controller/internal/datastore/datastore.go]: Or, after talking to @markdryan if you're just wanting to block then AddSubnet() can just return a channel, that might already be closed and <- on that. You can then use the same channel for all requests into AddSubnet()

File: ciao-controller/cnci.go [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+
+	cnci.ID = ""
+
+	cnci.transitionState(failed)
+
+	return nil
+}
+
+// WaitForActive will block until a cnci for a subnet is active
+func (c *CNCIManager) WaitForActive(subnet int) error {
```
Comment (rbradford) [inline ciao-controller/cnci.go]: I think this function can go away if the same channel is used to signify that the initial cnci launch vs a later "wait"

File: ciao-controller/client.go [hunk 5]
```diff
@@ -158,10 +175,33 @@ func (client *ssntpClient) instanceAdded(payload []byte) {
...
+	i, err := client.ctl.ds.GetInstance(newCNCI.InstanceUUID)
 	if err != nil {
-		glog.Warningf("Error adding CNCI IP to datastore: %v", err)
+		glog.Warningf("Error getting instance: %v", err)
+		return
+	}
+
+	// update the instance mac, ip, and subnet (TBD)
+	i.IPAddress = newCNCI.ConcentratorIP
+	i.MACAddress = newCNCI.ConcentratorMAC
```
Comment (mcastelino) [inline ciao-controller/client.go]: @kaccardi will there be another PR for CNCI to subnet tracking

File: ciao-controller/instance.go [hunk 6]
```diff
@@ -93,6 +95,11 @@ func newInstance(ctl *controller, tenantID string, workload *types.Workload,
 		MACAddress: config.mac,
 		CreateTime: time.Now(),
 		Name:       name,
+		StateLock:  &sync.RWMutex{},
+	}
+
+	if subnet != "" {
```
Comment (mcastelino) [inline ciao-controller/instance.go]: when will subnet be ""

File: ciao-controller/instance.go [hunk 7]
```diff
@@ -285,6 +321,91 @@ func controllerStorageResourceFromPayload(volume payloads.StorageResource) (s ty
...
+	// send in CIDR notation?
+	networking.PrivateIP = ipAddress.String()
+	mask := net.IPv4Mask(255, 255, 255, 0)
+	ipnet := net.IPNet{
+		IP:   ipAddress.Mask(mask),
+		Mask: mask,
+	}
+	networking.Subnet = ipnet.String()
+
+	cnciInstance, err := tenant.CNCIctrl.GetSubnetCNCI(networking.Subnet)
```
Comment (mcastelino) [inline ciao-controller/instance.go]: when will the CNCI get created when you end up on a different subnet

File: ciao-controller/cnci.go [hunk 8]
```diff
@@ -0,0 +1,459 @@
...
+// CNCIManager is a structure which defines a manager for CNCI instances
+// TBD: managing multiple CNCI instances.
+type CNCIManager struct {
+	tenant string
+	ctrl   *controller
+
+	// there's no reason to have separate lock for each map.
+	cnciLock *sync.RWMutex
+	cncis    map[string]*CNCI
+	subnets  map[int]*CNCI
```
Comment (mcastelino) [inline ciao-controller/cnci.go]: may be helpful to indicate what the map keys are in each case.

File: ciao-controller/cnci.go [hunk 9]
```diff
@@ -0,0 +1,459 @@
...
+	cnci.transitionState(failed)
+
+	// we should probably not do this, and instead we should
+	// delete from the map?
+	cnci.instance = nil
+
+	return nil
+}
+
+func (c *CNCIManager) waitForActive(subnet int) error {
```
Comment (rbradford) [inline ciao-controller/cnci.go]: Please rethink your use of channels. To do the multiple reader on the same channel you have two options:

a.) Close the channel. All readers will unblock and move forward. Removing the need for all the polling stuff below. We have code just like this in Ciao already at https://github.com/01org/ciao/blob/master/ciao-controller/command.go#L140

Obviously by closing a channel you can only convey completion.

b.) If you want to send the information down the channel then store a slice of channels in the struct (protected by a lock) and have everyone who wants to listen put their channel in the struct and then they can just wait for the answer on their own channel.

c.) Or using a condition variable: https://golang.org/pkg/sync/#Cond
No need for polling in either case.

File: ciao-controller/cnci.go [hunk 10]
```diff
@@ -0,0 +1,459 @@
...
+		cnci.instance = i
+
+		// convert cnci instance string to int for map
+		subnetInt, err := subnetStringToInt(i.Subnet)
+		if err != nil {
+			return nil, err
+		}
+
+		cnci.subnet = subnetInt
+
```
Comment (rbradford) [inline ciao-controller/cnci.go]: Some extra whitespace here and below.

File: ciao-controller/compute_test.go [hunk 11]
```diff
@@ -862,7 +862,8 @@ func testListNodeServers(t *testing.T, httpExpectedStatus int, validToken bool)
 		}
 
 		if result.TotalServers != len(instances) {
-			t.Fatal("Incorrect number of servers")
+			fmt.Printf("result: %v\n", result)
```
Comment (rbradford) [inline ciao-controller/compute_test.go]: Left over debugging?

File: ciao-controller/internal/datastore/datastore.go [hunk 12]
```diff
@@ -681,11 +562,52 @@ func (ds *Datastore) ReleaseTenantIP(tenantID string, ip string) error {
...
+			ds.tenantsLock.RUnlock()
+
+			if removeSubnet {
+				// handle error?
+				err := ds.tenants[tenantID].CNCIctrl.RemoveSubnet(i)
+				if err != nil {
+					glog.Warningf("Error removing subnet: (%v)\n", err)
+				}
+			}
+		}()
```
Comment (rbradford) [inline ciao-controller/internal/datastore/datastore.go]: In general when you start a goroutine you need to handle tidying up (especially for a long lived one). For this one, create a context.WithCancel() and save the cancel function into a map on the datastore (or in the tenant struct.)

In the goroutine select {} on both a time.After and the Done() channel on the context. On the former do your deletion on the latter just return and terminate the goroutine.

In the datastore shutdown call all the cancel functions  which will safely terminate all the running goroutines. You can also call the cancel if you get a new instance for a pending closed subnet removing the need to even check if its in use.

File: ciao-controller/client.go (hunk 5 above)
Comment (kaccardi) [inline ciao-controller/client.go]: This PR includes cnci -> subnet mapping.

File: ciao-controller/client.go [hunk 13]
```diff
@@ -114,26 +109,48 @@ func (client *ssntpClient) releaseResources(instanceID string) error {
 	return nil
 }
 
-func (client *ssntpClient) removeInstance(instanceID string) {
-	err := client.releaseResources(instanceID)
+func (client *ssntpClient) removeInstance(i *types.Instance) {
```
Comment (kaccardi) [inline ciao-controller/client.go]: @rbradford I reviewed the change again and think that this change was made in an earlier revision of the patch, and is no longer relevant. I'll revert it back to the original implementation.

Comment (kaccardi) [inline ciao-controller/client.go]: possibly I was also trying to avoid multiple calls to GetInstance in the other routines.

File: ciao-controller/instance.go (hunk 6 above)
Comment (kaccardi) [inline ciao-controller/instance.go]: whenever this is a standard tenant instance and not a CNCI.

File: ciao-controller/instance.go (hunk 7 above)
Comment (kaccardi) [inline ciao-controller/instance.go]: the CNCI is created when we allocate the tenant IP address down in the datastore. It will know whether this is a new/different subnet at that time and do the launch if needed.

File: ciao-controller/cnci.go (hunk 8 above)
Comment (kaccardi) [inline ciao-controller/cnci.go]: ok, I'll improve the comments.

File: ciao-controller/client.go [hunk 14]
```diff
@@ -114,26 +109,48 @@ func (client *ssntpClient) releaseResources(instanceID string) error {
...
+	tenant, err := client.ctl.ds.GetTenant(i.TenantID)
+	if err != nil {
+		glog.Warningf("Error retrieving tenant %v", err)
+		return
+	}
+
+	err = client.releaseResources(i)
 	if err != nil {
 		glog.Warningf("Error when releasing resources for deleted instance: %v", err)
```
Comment (kaccardi) [inline ciao-controller/client.go]: @rbradford going back and reviewing the original code here, it looks like it was intentionally not returning in order to attempt to try to complete as much of the removal as could be completed. I think that makes sense to keep.
